gen/internals.py

```diff
@@ -184,51 +221,203 @@ def add_entry(self, entry, replace_existing):
+    """
+
+    def __init__(self):
+        self._finalized = False
+
+    def __missing__(self, key):
+        assert not self._finalized, "No missing keys should be accessed"
+
+        value = self[key] = Resolvable(key)
+        return value
+
+    def finalize(self):
+        assert not self._finalized
+        self._finalized = True
+
+
+class Validator:
+    """Holds a collection of validate functions, and can be asked to call them"""
+
+    def __init__(self, validate_functions, targets):
+        # Note: targets must be passed in and inspected here, since the validate_functions that a
+        # target yields can't be inspected for the parameter name. To get around this yield_validates
+        # returns a two-tuple of the name and a callable.
 
         # Re-arrange the validation functions so we can more easily access them by
         # argument name.
         self._validate_by_arg = dict()
         self._multi_arg_validate = dict()
 
-        for fn in validate_fns:
-            parameters = get_function_parameters(fn)
+        for function in validate_functions:
+            parameters = get_function_parameters(function)
             # Could build up the single and multi parameter validation function maps in the same
             # thing but the timing / handling of when and how we run single vs. multi-parameter
             # validation functions is fairly different, the extra bit here simplifies the later code.
             if len(parameters) == 1:
-                self._validate_by_arg[parameters.pop()] = fn
-                assert not parameters
+                self._validate_by_arg.setdefault(parameters.pop(), list()).append(function)
             else:
-                self._multi_arg_validate[frozenset(parameters)] = fn
+                self._multi_arg_validate.setdefault(frozenset(parameters), list()).append(function)
+
+        for target in targets:
+            for parameter, function in target.yield_validates():
+                self._validate_by_arg.setdefault(parameter, list()).append(function)
+
+    def validate_single(self, name: str, value: str):
+        """Calls all validate functions which validate the given parameter name
+
+        The validate functions will raise an AssertionError which should be caught by the caller
+        when validation fails.
+        """
+        validate_fns = self._validate_by_arg.get(name)
+        if validate_fns is not None:
+            for validate_fn in validate_fns:
+                validate_fn(value)
+
+    # TODO(cmaloney): The distance between the validate_single and multi_arg_validate interface,
+    # while necessary for efficient functioning currently, is showing that there is tension between
+    # how / when each is run. Moving the multi-argument validation to as soon as possible after an
+    # argument set is finalized rather than one big pass at the end would likely make it much
+    # cleaner.
```

Comment:
The problem with that is we need to do single-argument validate functions as soon as possible after a value is calculated / before we .resolve() a Resolvable with it. 

That's necessary, so that invalid values never are accessible for other calculation functions or validation functions, where their invalid / incorrect value may cause issues (Code that takes resolvers among other things should be able to assume resolvers is a json list).

Ideally, as soon as we had all the arguments for a multi-arg validate, we would check that as well. Unfortunately, that is more complicated code. The simpler code to do it for multi-argument validates would be to basically say
 - Every time an argument has a new value tell validator about it
 - Have validator iterate through every validation function which takes the new argument as one of it's parameters and if all the parameters are set, then validate it.
 - Call the validate functions, sorting the single-argument validate functions before the multi-argument validate functions with the same signature.

Doing that though I think is out of scope for just what this code is doing, definitely adds more complexity / logic. The current evaluation method is not that complicated: Single arg ones are validated instantly, which is pretty simple to do. If a single-arg validate finds a value to be invalid, the argument won't show up in the multi_argument_validate argument dict as resolved and hence multi-arg validates won't get called

Multi-arg ones are validated at the end of resolution, where we simply iterate once over all the multi-argument validate functions, checking if the validate function arguments are a subset of all the arguments, and if so calling it and yielding out the errors.